File: Assets/Game/Scripts/UI/GamePlayUIControl.lua
local GameManager = Ballance2.Services.GameManager
local GameUIManager = GameManager.GetSystemService('GameUIManager') ---@type GameUIManager
local GameSettingsManager = Ballance2.Services.GameSettingsManager
local Log = Ballance2.Log
local CloneUtils = Ballance2.Utils.CloneUtils
local Image = UnityEngine.UI.Image
local Color = UnityEngine.Color
local Vector2 = UnityEngine.Vector2
local InputField = UnityEngine.UI.InputField
local Yield = UnityEngine.Yield
local Mathf = UnityEngine.Mathf
local Time = UnityEngine.Time
local WaitForSeconds = UnityEngine.WaitForSeconds
local PlayerPrefs = UnityEngine.PlayerPrefs

---创建主游戏菜单UI
---@param package GamePackage
function CreateGamePlayUI(package)

  local PageGamePause = GameUIManager:RegisterPage('PageGamePause', 'PageCommon')
  local PageGameFail = GameUIManager:RegisterPage('PageGameFail', 'PageCommon')
  local PageGameQuitAsk = GameUIManager:RegisterPage('PageGameQuitAsk', 'PageCommon')
  local PageGameRestartAsk = GameUIManager:RegisterPage('PageGameRestartAsk', 'PageCommon')
  local PageGameWinRestartAsk = GameUIManager:RegisterPage('PageGameWinRestartAsk', 'PageCommon')
  local PageGameWin = GameUIManager:RegisterPage('PageGameWin', 'PageCommon')
  local PageEndScore = GameUIManager:RegisterPage('PageEndScore', 'PageTransparent')
  local PageHighscoreEntry = GameUIManager:RegisterPage('PageHighscoreEntry', 'PageCommon')
  local PageGamePreviewPause = GameUIManager:RegisterPage('PageGamePreviewPause', 'PageCommon')
  local PageGamePreviewQuitAsk = GameUIManager:RegisterPage('PageGamePreviewQuitAsk', 'PageCommon')
  
  coroutine.resume(coroutine.create(function ()
    
    PageGamePause:CreateContent(package)
    PageGameQuitAsk:CreateContent(package)
    PageGameRestartAsk:CreateContent(package)
    Yield(WaitForSeconds(0.1))
    PageGamePause:CreateContent(package)
    PageGamePause.CanEscBack = false
    PageGameWin:CreateContent(package)
    Yield(WaitForSeconds(0.1))
    PageGameWinRestartAsk:CreateContent(package)
    PageGameWin.CanEscBack = false
    PageEndScore:CreateContent(package)
    PageEndScore.CanEscBack = false
    Yield(WaitForSeconds(0.1))
    PageHighscoreEntry:CreateContent(package)
    PageHighscoreEntry.CanEscBack = false
    PageGameFail:CreateContent(package)
    PageGameFail.CanEscBack = false
    Yield(WaitForSeconds(0.1))
    PageGamePreviewPause:CreateContent(package)
    PageGamePreviewQuitAsk:CreateContent(package)

    MessageCenter:SubscribeEvent('BtnGameHomeClick', function () GamePlay.GamePlayManager:QuitLevel() end)
    MessageCenter:SubscribeEvent('BtnNextLevellick', function () GamePlay.GamePlayManager:NextLevel() end)
    MessageCenter:SubscribeEvent('BtnGameRestartClick', function () GameUIManager:GoPage('PageGameRestartAsk') end)
    MessageCenter:SubscribeEvent('BtnGameWinRestartClick', function () GameUIManager:GoPage('PageGameWinRestartAsk') end)
    MessageCenter:SubscribeEvent('BtnGameQuitClick', function () GameUIManager:GoPage('PageGameQuitAsk') end)
    MessageCenter:SubscribeEvent('BtnGamePreviewQuitClick', function () GameUIManager:GoPage('PageGamePreviewQuitAsk') end)
    MessageCenter:SubscribeEvent('BtnPauseSettingsClick', function () GameUIManager:GoPage('PageSettingsInGame') end)
    MessageCenter:SubscribeEvent('BtnGameFailRestartClick', function ()
      GameUIManager:HideCurrentPage()
      Game.GamePlay.GamePlayManager:RestartLevel()
    end)
    MessageCenter:SubscribeEvent('BtnGameQuitSureClick', function ()
      GameUIManager:HideCurrentPage()
      Game.GamePlay.GamePlayManager:QuitLevel()
    end)
    MessageCenter:SubscribeEvent('BtnGameFailQuitClick', function ()
      GameUIManager:HideCurrentPage()
      Game.GamePlay.GamePlayManager:QuitLevel()
    end)
    MessageCenter:SubscribeEvent('BtnResumeClick', function () Game.GamePlay.GamePlayManager:ResumeLevel() end)
    MessageCenter:SubscribeEvent('BtnPreviewResumeClick', function () Game.GamePlay.GamePreviewManager:ResumeLevel() end)
    MessageCenter:SubscribeEvent('BtnGamePreviewQuitSureClick', function ()
      GameUIManager:HideCurrentPage()
      Game.GamePlay.GamePreviewManager:QuitLevel()
    end)

    --高分默认数据
    local HighscoreEntryName = PageHighscoreEntry.Content:Find('InputField'):GetComponent(InputField) ---@type InputField
    HighscoreEntryName.text = PlayerPrefs.GetString('LastEnterHighscoreEntry', 'NAME')

    --过关之后的下一关按扭
    local ButtonNext = PageGameWin.Content:Find('ButtonNext').gameObject
    PageGameWin.OnShow = function ()
      ButtonNext:SetActive(GamePlay.GamePlayManager.NextLevelName ~= '')
    end

    MessageCenter:SubscribeEvent('BtnHighscrollEnterClick', function () 
      PlayerPrefs.SetString('LastEnterHighscoreEntry', HighscoreEntryName.text)
      GameUI.WinScoreUIControl:SaveHighscore(HighscoreEntryName.text)

      --当前有新的高分，跳转到高分页，否则直接进入下一关菜单
      GameUIManager:GoPage('PageGameWin') 
      if GameUI.WinScoreUIControl._ThisTimeHasNewHighscore then
        LuaTimer.Add(200, function ()
          GameUIManager:GoPage('PageHighscore')
          --高分页加载当前关卡数据
          GameUI.HighscoreUIControl:LoadLevelData(GamePlay.GamePlayManager.CurrentLevelName)
        end)
      end
    end)
  end))
end

---主游戏菜单控制器类
---@class GamePlayUIControl : GameLuaObjectHostClass
---@field _ScoreBoardActive Image
---@field _ScoreText Text
---@field _LifeBoardLeftBaffle Image
---@field _LifeBoardBallPrefab GameObject
---@field _LifeBoardBallInfPrefab GameObject
---@field _LifeBalls RectTransform
---@field _DebugStats GuiStats
---@field _DebugStatValues GuiStatsValue[]
GamePlayUIControl = ClassicObject:extend()

function GamePlayUIControl:new() 
  self._CurrentShowLifeBallCount = 0
  self._CurrentMoveBaffleTick = 0.3
  self._MoveBaffleSec = 0.3
  self._CurrentMoveBaffleStart = 0
  self._CurrentMoveBaffleTarget = 0
  self._CurrentMobileKeyPad = nil ---@type GameObject
  self._CurrentMobileKeyPadShow = false
  self._DebugStatValues = {}
end
function GamePlayUIControl:Start() 
  self._ScoreBoardActive.gameObject:SetActive(false)

  --创建调试信息
  if BALLANCE_DEBUG then
    self._DebugStatValues['CurrentBall'] = self._DebugStats:AddStat('CurrentBall')
    self._DebugStatValues['CurrentStatus'] = self._DebugStats:AddStat('CurrentStatus')
    self._DebugStatValues['Position'] = self._DebugStats:AddStat('Position')
    self._DebugStatValues['Rotation'] = self._DebugStats:AddStat('Rotation')
    self._DebugStatValues['CamDirection'] = self._DebugStats:AddStat('CamDirection')
    self._DebugStatValues['CamState'] = self._DebugStats:AddStat('CamState')
    self._DebugStatValues['Velocity'] = self._DebugStats:AddStat('Velocity')
    self._DebugStatValues['PushValue'] = self._DebugStats:AddStat('PushValue')
    self._DebugStatValues['Sector'] = self._DebugStats:AddStat('Sector')
    self._DebugStatValues['Moduls'] = self._DebugStats:AddStat('Moduls')
  end

  --手机端还需要创建键盘
  if UNITY_ANDROID or UNITY_IOS then
    self:ReBuildMobileKeyPad()
    self._CurrentMobileKeyPadShow = false
    GamePlay.GamePlayManager.EventStart:On(function ()
      self._CurrentMobileKeyPadShow = true
      if self._CurrentMobileKeyPad then
        self._CurrentMobileKeyPad.gameObject:SetActive(true)
      end
    end)
    GamePlay.GamePlayManager.EventQuit:On(function ()
      self._CurrentMobileKeyPadShow = false
      if self._CurrentMobileKeyPad then
        self._CurrentMobileKeyPad.gameObject:SetActive(false)
      end
    end)
  end

  GameUI.GamePlayUI = self
end
function GamePlayUIControl:Update()
  if self._CurrentMoveBaffleTick < self._MoveBaffleSec then
    self._CurrentMoveBaffleTick = self._CurrentMoveBaffleTick + Time.deltaTime
    self._LifeBoardLeftBaffle.rectTransform.anchoredPosition = Vector2(
      Mathf.Lerp(self._CurrentMoveBaffleStart, self._CurrentMoveBaffleTarget, self._CurrentMoveBaffleTick / self._MoveBaffleSec),
      0
    )
  end
end
function GamePlayUIControl:OnDestroy()
  --销毁键盘
  if self._CurrentMobileKeyPad and not Slua.IsNull(self._CurrentMobileKeyPad) then
    UnityEngine.Object.Destroy(self._CurrentMobileKeyPad.gameObject)
    self._CurrentMobileKeyPad = nil
  end
end

---创建手机端键盘
function GamePlayUIControl:ReBuildMobileKeyPad() 
  --销毁键盘
  if self._CurrentMobileKeyPad and not Slua.IsNull(self._CurrentMobileKeyPad) then
    UnityEngine.Object.Destroy(self._CurrentMobileKeyPad.gameObject)
    self._CurrentMobileKeyPad = nil
  end

  --读取键盘设置
  local settings = GameSettingsManager.GetSettings('core')
  local controlKeypadSettting = settings:GetString('control.keypad', 'BaseLeft')

  local keyPad = KeypadUIManager.GetKeypad(controlKeypadSettting)
  if keyPad == nil then
    Log.E('GamePlayUIControl', 'Keypad in setting \''..controlKeypadSettting..'\' not found, use default keypad insted.')
    keyPad = KeypadUIManager.GetKeypad('BaseLeft')
  end

  --创建键盘
  self._CurrentMobileKeyPad = GameUIManager:InitViewToCanvas(keyPad.prefrab, 'GameMobileKeypad', false)

  if not self._CurrentMobileKeyPadShow then
    self._CurrentMobileKeyPad.gameObject:SetActive(false)
  end
end

---闪烁分数面板
function GamePlayUIControl:TwinklePoint() 
  self._ScoreBoardActive.gameObject:SetActive(true)
  GameUIManager.UIFadeManager:AddFadeOut(self._ScoreBoardActive, 1, true)
end
---设置分数面板文字
---@param score number|string 分数面板文字
function GamePlayUIControl:SetPointText(score) 
  if type(score) == "number" then
    self._ScoreText.text = string.format("%d", score)
  elseif type(score) == "string" then 
    self._ScoreText.text = score
  end
end

---当前显示的生命球数 +1
function GamePlayUIControl:AddLifeBall()
  if self._CurrentShowLifeBallCount ~= -1 then
    self._CurrentShowLifeBallCount = self._CurrentShowLifeBallCount + 1

    local ball = CloneUtils.CloneNewObjectWithParent(self._LifeBoardBallPrefab, self._LifeBalls):GetComponent(Image) ---@type Image
    ball.rectTransform:SetAsFirstSibling()
    ball.color = Color(1,1,1,0)
    coroutine.resume(coroutine.create(function()
      self:_MoveLifeLeftBaffle()
      Yield(WaitForSeconds(0.4))
      GameUIManager.UIFadeManager:AddFadeIn(ball, 0.4)
    end))
  end
end
---当前显示的生命球数 -1
function GamePlayUIControl:RemoveLifeBall()
  if self._CurrentShowLifeBallCount > 0 then
    self._CurrentShowLifeBallCount = self._CurrentShowLifeBallCount - 1

    local ball = self._LifeBalls:GetChild(self._LifeBalls.childCount - 1 - self._CurrentShowLifeBallCount)
    GameUIManager.UIFadeManager:AddFadeOut(ball:GetComponent(Image), 0.4, true)
    coroutine.resume(coroutine.create(function()
      Yield(WaitForSeconds(0.4))
      self:_MoveLifeLeftBaffle()
      UnityEngine.Object.Destroy(ball.gameObject)
    end))
  end
end
---直接设置当前显示的生命球数（无动画效果）
---@param count any
function GamePlayUIControl:SetLifeBallCount(count)
  if count ~= self._CurrentShowLifeBallCount then
    self._CurrentShowLifeBallCount = count
    if count == -1 then
      --负1就显示无限大图标
      for i = self._LifeBalls.childCount - 1, 0, -1 do
        UnityEngine.Object.Destroy(self._LifeBalls:GetChild(i).gameObject)
      end
      CloneUtils.CloneNewObjectWithParent(self._LifeBoardBallInfPrefab, self._LifeBalls)
      self:_MoveLifeLeftBaffle()
    else
      if self._LifeBalls.childCount > count then
        --显示数量大于目标，删除多余的
        for i = self._LifeBalls.childCount - 1, count, -1 do
          UnityEngine.Object.Destroy(self._LifeBalls:GetChild(i).gameObject)
        end
        self:_MoveLifeLeftBaffle()
      elseif self._LifeBalls.childCount < count then
        --显示数量小于目标，添加
        for i = 1, count - self._LifeBalls.childCount, 1 do
          CloneUtils.CloneNewObjectWithParent(self._LifeBoardBallPrefab, self._LifeBalls)
        end
        self:_MoveLifeLeftBaffle()
      end
    end
  end
end

function GamePlayUIControl:_MoveLifeLeftBaffle() 
  if self._CurrentShowLifeBallCount < 0 then
    self._CurrentMoveBaffleTarget =  -27
  else
    self._CurrentMoveBaffleTarget =  -(self._CurrentShowLifeBallCount * 27)
  end
  self._CurrentMoveBaffleTick = 0
  self._CurrentMoveBaffleStart = self._LifeBoardLeftBaffle.rectTransform.anchoredPosition.x
end

function CreateClass:GamePlayUIControl() 
  return GamePlayUIControl()
end

File: Assets/Game/Menu/Scripts/Game/GamePlayUIControl.cs
using System.Collections.Generic;
using Ballance2.Base;
using Ballance2.Game;
using Ballance2.Game.GamePlay;
using Ballance2.Services;
using Ballance2.Utils;
using UnityEngine;
using UnityEngine.UI;

namespace Ballance2.Menu
{
  /// <summary>
  /// 主游戏菜单控制器类
  /// </summary>
  public class GamePlayUIControl : GameSingletonBehavior<GamePlayUIControl> 
  {
    public Image _ScoreBoardActive ;
    public Text _ScoreText;
    public Image _LifeBoardLeftBaffle;
    public GameObject _LifeBoardBallPrefab;
    public GameObject _LifeBoardBallInfPrefab;
    public RectTransform _LifeBalls;
    public GameObject _TextDebugMode;
    public GuiStats GuiStats;

    private int _CurrentShowLifeBallCount = 0;
    private float _CurrentMoveBaffleTick = 0.3f;
    private float _MoveBaffleSec = 0.3f;
    private float _CurrentMoveBaffleStart = 0;
    private float _CurrentMoveBaffleTarget = 0;
    private RectTransform _CurrentMobileKeyPad = null;
    private bool _CurrentMobileKeyPadShow = false;

    public Dictionary<string, GuiStatsValue> DebugStatValues = new Dictionary<string, GuiStatsValue>();

    private void Start() {
      _ScoreBoardActive.gameObject.SetActive(false);

        //手机端还需要创建键盘
        #if (UNITY_ANDROID || UNITY_IOS)
        //当前设备存在触摸屏时，才允许显示屏幕键盘
        if (UnityEngine.InputSystem.Touchscreen.current != null)
        {
            _CurrentMobileKeyPadShow = true;
            ReBuildMobileKeyPad();
            GamePlayManager.Instance.EventBeforeStart.On((evt) =>
            {
                _CurrentMobileKeyPadShow = true;
                if (_CurrentMobileKeyPad != null)
                    _CurrentMobileKeyPad.gameObject.SetActive(true);
            });
            GamePlayManager.Instance.EventQuit.On((evt) =>
            {
                _CurrentMobileKeyPadShow = false;
                if (_CurrentMobileKeyPad != null)
                    _CurrentMobileKeyPad.gameObject.SetActive(false);
            });
        }
      #endif

      //创建调试信息
      if (GameManager.DebugMode) {
        DebugStatValues["CurrentBall"] = GuiStats.AddStat("CurrentBall");
        DebugStatValues["CurrentStatus"] = GuiStats.AddStat("CurrentStatus");
        DebugStatValues["Position"] = GuiStats.AddStat("Position");
        DebugStatValues["Rotation"] = GuiStats.AddStat("Rotation");
        DebugStatValues["PhysicsState"] = GuiStats.AddStat("PhysicsState");
        DebugStatValues["CamDirection"] = GuiStats.AddStat("CamDirection");
        DebugStatValues["CamState"] = GuiStats.AddStat("CamState");
        DebugStatValues["Velocity"] = GuiStats.AddStat("Velocity");
        DebugStatValues["PushValue"] = GuiStats.AddStat("PushValue");
        DebugStatValues["Sector"] = GuiStats.AddStat("Sector");
        DebugStatValues["Moduls"] = GuiStats.AddStat("Moduls");
        DebugStatValues["PhysicsTime"] = GuiStats.AddStat("PhysicsTime");
        DebugStatValues["PhysicsObjects"] = GuiStats.AddStat("PhysicsObjects");
        _TextDebugMode.SetActive(true);
      } else {
        _TextDebugMode.SetActive(false);
      }
    }

    private void Update()
    {
      if (_CurrentMoveBaffleTick < _MoveBaffleSec)
      {
        _CurrentMoveBaffleTick = _CurrentMoveBaffleTick + Time.deltaTime;
        _LifeBoardLeftBaffle.rectTransform.anchoredPosition = new Vector2(
          Mathf.Lerp(_CurrentMoveBaffleStart, _CurrentMoveBaffleTarget, _CurrentMoveBaffleTick / _MoveBaffleSec),
          0
        );
      }
    }

    protected override void OnDestroy()
    {
      base.OnDestroy();
      this.DestroyMobileKeyPad();
    }

    private void DestroyMobileKeyPad() {
      //销毁键盘
      if (_CurrentMobileKeyPad != null) {
        Destroy(_CurrentMobileKeyPad.gameObject);
        _CurrentMobileKeyPad = null;
      }
    }
    internal void ReBuildMobileKeyPad() {
      //销毁键盘
      DestroyMobileKeyPad();
      //读取键盘设置
      var settings = GameSettingsManager.GetSettings(GamePackageManager.SYSTEM_PACKAGE_NAME);
      var controlKeypadSettting = settings.GetString(SettingConstants.SettingsControlKeypad);

      KeypadUIManager.KeypadUIInfo keyPad;
      if (!KeypadUIManager.GetKeypadRegistered(controlKeypadSettting)) 
      {
        Log.E("GamePlayUIControl", "Keypad in setting \"" + controlKeypadSettting + "\" not found, use default keypad insted.");
        keyPad = KeypadUIManager.GetKeypad("BaseCenter");
      } 
      else 
      {
        keyPad = KeypadUIManager.GetKeypad(controlKeypadSettting);
      }

      //创建键盘
      _CurrentMobileKeyPad = GameUIManager.Instance.InitViewToCanvas(keyPad.prefab, "GameMobileKeypad", false);
      //当屏幕键盘不显示
      if (!_CurrentMobileKeyPadShow)
        _CurrentMobileKeyPad.gameObject.SetActive(false);
    }

    /// <summary>
    /// 闪烁分数面板
    /// </summary>
    public void TwinklePoint() 
    {
      _ScoreBoardActive.gameObject.SetActive(true);
      GameUIManager.Instance.UIFadeManager.AddFadeOut(_ScoreBoardActive, 1, true);
    }
    /// <summary>
    /// 设置分数面板文字
    /// </summary>
    /// <param name="score">分数</param>
    public void SetPointText(int score) 
    {
      _ScoreText.text = score.ToString();
    }
    /// <summary>
    /// 当前显示的生命球数 +1
    /// </summary>
    public void AddLifeBall()
    {
      if (_CurrentShowLifeBallCount != -1) {
        _CurrentShowLifeBallCount++;

        var ball = CloneUtils.CloneNewObjectWithParent(_LifeBoardBallPrefab, _LifeBalls).GetComponent<Image>();
        ball.rectTransform.SetAsFirstSibling();
        ball.color = new Color(1,1,1,0);
        _MoveLifeLeftBaffle();
        GameTimer.Delay(0.4f, () => {
          GameUIManager.Instance.UIFadeManager.AddFadeIn(ball, 0.4f);
        });
      }
    }
    /// <summary>
    /// 当前显示的生命球数 -1
    /// </summary>
    public void RemoveLifeBall()
    {
      if (_CurrentShowLifeBallCount <= -1) 
        return;
      if (_CurrentShowLifeBallCount > 0) {
        _CurrentShowLifeBallCount--;
        var ballIndex = _LifeBalls.childCount - 1 - _CurrentShowLifeBallCount;
        if (ballIndex < 0) ballIndex = 0;
        if (ballIndex == 0 && _LifeBalls.childCount == 0) {
          _MoveLifeLeftBaffle();
          return;
        }

        var ball = _LifeBalls.GetChild(ballIndex);
        GameUIManager.Instance.UIFadeManager.AddFadeOut(ball.GetComponent<Image>(), 0.4f, true);
        GameTimer.Delay(0.4f, () => {
          _MoveLifeLeftBaffle();
          UnityEngine.Object.Destroy(ball.gameObject);
        });
      }
    }
    /// <summary>
    /// 直接设置当前显示的生命球数（无动画效果）
    /// </summary>
    /// <param name="count">数量</param>
    public void SetLifeBallCount(int count)
    {
      if (count != _CurrentShowLifeBallCount) {
        _CurrentShowLifeBallCount = count;
        if (count == -1) {
          //负1就显示无限大图标
          for (var i = _LifeBalls.childCount - 1; i >= 0; i--)
            Destroy(_LifeBalls.GetChild(i).gameObject);
          CloneUtils.CloneNewObjectWithParent(_LifeBoardBallInfPrefab, _LifeBalls);
          _MoveLifeLeftBaffle();
        }
        else {
          if (_LifeBalls.childCount > count) {
            //显示数量大于目标，删除多余的
            for (var i = _LifeBalls.childCount - 1; i >= count; i--)
              UnityEngine.Object.Destroy(_LifeBalls.GetChild(i).gameObject);
            _MoveLifeLeftBaffle();
          }
          else if (_LifeBalls.childCount < count) {
            //显示数量小于目标，添加
            for (var i = count - _LifeBalls.childCount; i > 0; i--)
              CloneUtils.CloneNewObjectWithParent(_LifeBoardBallPrefab, _LifeBalls);
            _MoveLifeLeftBaffle();
          }
        }
      }
    }

    private void _MoveLifeLeftBaffle() 
    {
      if (_CurrentShowLifeBallCount < 0)
        _CurrentMoveBaffleTarget =  -27;
      else
        _CurrentMoveBaffleTarget =  -(_CurrentShowLifeBallCount * 27);
      _CurrentMoveBaffleTick = 0;
      _CurrentMoveBaffleStart = _LifeBoardLeftBaffle.rectTransform.anchoredPosition.x;
    }
  }
}